src/kudu/consensus/raft_consensus.cc

```diff
@@ -3297,6 +3331,280 @@ void RaftConsensus::SetLeaderDetectedCallback(LeaderDetectedCallback ldcb) {
+  return !request->proxy_dest_uuid().empty();
+}
+
+// Set an error and respond.
+// Stolen (mostly) from tablet_service.cc
+static void SetupErrorAndRespond(const Status& s,
+                                 ServerErrorPB::Code code,
+                                 ConsensusResponsePB* response,
+                                 rpc::RpcContext* context) {
+  // Generic "service unavailable" errors will cause the client to retry later.
+  if ((code == ServerErrorPB::UNKNOWN_ERROR /*||
+       code == TabletServerErrorPB::THROTTLED */) && s.IsServiceUnavailable()) {
+    context->RespondRpcFailure(rpc::ErrorStatusPB::ERROR_SERVER_TOO_BUSY, s);
+    return;
+  }
+
+  StatusToPB(s, response->mutable_error()->mutable_status());
+  response->mutable_error()->set_code(code);
+  context->RespondNoCache();
+}
+
+// Respond with an error and return if 's' is not OK.
+#define RET_RESPOND_ERROR_NOT_OK(s) \
+  do { \
+    const kudu::Status& _s = (s); \
+    if (PREDICT_FALSE(!_s.ok())) { \
+      SetupErrorAndRespond(_s, ServerErrorPB::UNKNOWN_ERROR, response, context); \
+      return; \
+    } \
+  } while (0)
+
+void RaftConsensus::HandleProxyRequest(const ConsensusRequestPB* request,
+                                       ConsensusResponsePB* response,
+                                       rpc::RpcContext* context) {
+  MonoDelta wal_wait_timeout = MonoDelta::FromMilliseconds(FLAGS_raft_log_cache_proxy_wait_time_ms);
+  MonoTime wal_wait_deadline = MonoTime::Now() + wal_wait_timeout;
+
+  // TODO(mpercy): Remove this config lookup when refactoring DRT to return a
+  // RaftPeerPB, which will prevent a validation race.
+  RaftConfigPB active_config;
+  {
+    // Snapshot the active Raft config so we know how to route proxied messages.
+    ThreadRestrictions::AssertWaitAllowed();
+    LockGuard l(lock_);
+    RET_RESPOND_ERROR_NOT_OK(CheckRunningUnlocked());
+    active_config = cmeta_->ActiveConfig();
+  }
+
+  // Initial implementation:
+  //
+  // Synchronously:
+  // 1. Validate that the request is addressed to the local node via 'proxy_dest_uuid'.
+  // 2. Reconstitute each message from the local cache.
+  //
+  // Asynchronously:
+  // 4. Deliver the reconstituted request directly to the remote (async).
+  // 5. Proxy the response from the remote back to the caller.
+
+  // Validate the request.
+  if (request->proxy_dest_uuid() != peer_uuid()) {
+    Status s = Status::InvalidArgument(Substitute("Wrong proxy destination UUID requested. "
+                                                  "Local UUID: $1. Requested UUID: $2",
+                                                  peer_uuid(), request->proxy_dest_uuid()));
+    LOG_WITH_PREFIX(WARNING) << s.ToString() << ": from " << context->requestor_string()
+                 << ": " << SecureShortDebugString(*request);
+    SetupErrorAndRespond(s, ServerErrorPB::WRONG_SERVER_UUID, response, context);
+    return;
+  }
+  if (request->dest_uuid() == peer_uuid()) {
```

Comment:
Is this a valid check. the proxy topology can change under your feet because of time lag between
dbstatus putting new config. This kind of message rejection can lead to availability loss.